fix: validate remote validator signatures

adria0pol opened the pull request:
## 🔄 Changes Summary
- At this moment, external validator signatures are not checked, since these signatures are packed and sent to the agglayer node, if any validator generates a bad signature the certificate will be rejected and aggkit are not going to be able to recover from this error.

Stefan-Ethernal committed:
fix: nightly e2e tests (#1132)

## 🔄 Changes Summary
Some e2e tests, that run exclusively in nightly builds are still
failing:
- disable zkevm-bridge-service, because for `internal-claims.bats`
tests, the bridge gets claimed, before the proof and the exit root gets
manipulated (which is expected to eventually fail)
- bump the e2e ref, that includes several fixes, described in
https://github.com/agglayer/e2e/pull/194

## ⚠️ Breaking Changes
N/A

## 📋 Config Updates
N/A

## ✅ Testing
- 🤖 **Automatic**: Nightly e2e tests
(https://github.com/agglayer/aggkit/actions/runs/18679998702)
- 🖱️ **Manual**: N/A

## 🐞 Issues
- Closes #1133 

## 🔗 Related PRs
https://github.com/agglayer/e2e/pull/194

## 📝 Notes
The only test that is still
[failing](https://github.com/agglayer/aggkit/actions/runs/18679998702/job/53259017930)
is the `Bridge message A → Bridge asset B → Claim message A → Claim
asset B` in the op-succinct stream, due to `InvalidSmtProof` execution
revert error.

adria0pol committed:
fix(aggsender): validate remote validator sigs

At this moment, this is not checked, since these signatures are packed
and sent to the agglayer node, if any validator generates a bad
signature the certificate will be rejected and aggkit are not going to
be able to recover from this error.

rachit77 committed:
feat: remove calldata from bridge service apis (#1131)

## 🔄 Changes Summary
- This PR removes `calldata` field from `bridge`, `token_mapping` and
`legacy_token_migration` table of bridge sevice

## ✅ Testing
- 🤖 **Automatic**: [Optional: Enumerate E2E tests]
- 🖱️ **Manual**: [Optional: Steps to verify]

## 🐞 Issues
- Closes #1121

Stefan-Ethernal committed:
test: remove fixed gas price when sending tx (#1142)

## 🔄 Changes Summary
Remove fixed gas price when running bats tests. The test `Test triple
claim internal calls -> 1 fail, 1 success and 1 fail` has failed when
was sending tx with fixed gas price, so the doubt is that this has
caused it
(https://github.com/agglayer/aggkit/actions/runs/18701381470/job/53330826031)

## ⚠️ Breaking Changes
N/A

## 📋 Config Updates
N/A

## ✅ Testing
- 🤖 **Automatic**: E2E tests are executed succesfully
- 🖱️ **Manual**: N/A

## 🐞 Issues
- Closes #1143 

## 🔗 Related PRs
https://github.com/agglayer/e2e/pull/197

## 📝 Notes
- [Optional: design decisions, tradeoffs, or TODOs]

adria0pol committed:
Fixed/added tests

goran-ethernal committed:
fix: cherry pick of PR 1145 - `imported-bridge-exit` commitment hash in optimistic mode (#1146)

## 🔄 Changes Summary
This PR is a cherry pick of #1145 

## ⚠️ Breaking Changes
NA

## 📋 Config Updates
NA

## ✅ Testing
- 🤖 **Automatic**: `aggkit` CI

Stefan-Ethernal commented on aggsender/validator/remote_validator.go:
```diff
@@ -76,6 +77,15 @@ func (v *RemoteValidator) ValidateAndSignCertificate(
 	lastL2BlockInCert uint64,
 ) ([]byte, error) {
+	// Get the hash of the certificate, fail fast if something is wrong in it
+
```
Nitpick, no need to have the empty line IMO
```suggestion
```

adria0pol replied:
Apply changes on @Stefan-Ethernal review

Stefan-Ethernal commented on aggsender/validator/remote_validator.go:
```diff
@@ -96,5 +106,19 @@ func (v *RemoteValidator) ValidateAndSignCertificate(
+
+	recoveredAddress := crypto.PubkeyToAddress(*recoveredPublicKey)
+	if v.address != recoveredAddress {
+		return nil, fmt.Errorf("error validating remote validator signature, mismatch expected:%v current:%v",
```
```suggestion
		return nil, fmt.Errorf("error validating remote validator signature, mismatch in addresses (expected: %s, recovered: %s"),
```

adria0pol replied:
Getting a different address doesn’t necessarily mean that the recovered address is the actual signer of the certificate, so I’m a bit hesitant to use the term “address” here.

Stefan-Ethernal commented on aggsender/validator/remote_validator.go:
```diff
@@ -76,6 +77,15 @@ func (v *RemoteValidator) ValidateAndSignCertificate(
+		return nil, fmt.Errorf("internal error getting certificate hash: %w", err)
+	}
+
+	// Request remote signature
```
Is this comment misplaced?
```suggestion
```

Stefan-Ethernal commented on aggsender/validator/remote_validator.go:
```diff
@@ -96,5 +106,19 @@ func (v *RemoteValidator) ValidateAndSignCertificate(
+	// Validate received signature
+	// We do not support ethereum legacy v+27 signatures
+
+	recoveredPublicKey, err := crypto.SigToPub(certificateHash[:], signature)
+	if err != nil {
+		return nil, fmt.Errorf("error validating remote validator signature: %w", err)
+	}
+
+	recoveredAddress := crypto.PubkeyToAddress(*recoveredPublicKey)
+	if v.address != recoveredAddress {
+		return nil, fmt.Errorf("error validating remote validator signature, mismatch expected:%v current:%v",
+			v.address, recoveredAddress)
+	}
```
Does it make sense to run this check immediately after we have calculated the certificate hash, WDYT?
https://github.com/agglayer/aggkit/blob/a1ed47ce82dbc3f13155acaab0bb9347f00f8e53/aggsender/validator/remote_validator.go#L82-L85

adria0pol replied:
~Agree!~ 

update:

I execute it here because if it fails we do not have to do an external call for verifying the certificate. For sure what you say favours the readeability via code blocks.

adria0pol committed:
Apply changes on @Stefan-Ethernal review

Co-authored-by: Stefan Negovanović <93934272+Stefan-Ethernal@users.noreply.github.com>

adria0pol committed:
make lint happy

tnobayashi committed:
fix: docker arm image (#1151)

## 🔄 Changes Summary
### Context
- `amd-runner-2204` (**x86_64**) builds for both `linux/amd64` and
`linux/arm64`
- The `FROM` instruction is pinned to the native platform of the builder
by using `--platform=${BUILDPLATFORM}` in the Dockerfile
([ref](https://docs.docker.com/build/building/multi-platform/#cross-compilation))
- Additionally, `ARCH := $(shell arch)` in the Makefile returns
**x86_64**, not the target platform architecture

```
$ docker run --rm -it ghcr.io/agglayer/aggkit:0.7.1-rc1@sha256:77bbffc1e944ae0979a99a38978bf77d31f7ced53569f4098cc163ad86f2137c
rosetta error: failed to open elf at /lib/ld-musl-x86_64.so.1
```

### Approach
- Build native images using `amd-runner-2204` and `arm-runner-2204`


## ⚠️ Breaking Changes
- NA
## 📋 Config Updates
- NA

## ✅ Testing
- Manually triggered the [Build and Push Docker
Image](https://github.com/agglayer/aggkit/actions/runs/18760171063)
workflow

```
docker run --rm -it ghcr.io/agglayer/aggkit:tnobayashi-docker-arm_2025_10_23_19_57_9f47c74@sha256:0f7e54e1ceacc876c20c07f79d2f4ac1a59c679c613cb0611b25750f282e1db7
NAME:
   aggkit - A new cli application

USAGE:
   aggkit [global options] command [command options]

COMMANDS:
   version  Application version and build
   run      Run the aggkit client
   help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --help, -h  show help
```

Stefan-Ethernal committed:
refactor: bridge service metrics to prometheus (#1156)

## 🔄 Changes Summary
Refactor the bridge service metrics to rely on Prometheus metrics. All
the metrics exposed by the bridge service are prefixed with `bridge_`.

## ⚠️ Breaking Changes
N/A

## 📋 Config Updates
N/A

## ✅ Testing
- 🤖 **Automatic**: N/A
- 🖱️ **Manual**: N/A

## 🐞 Issues
- Closes #519

## 🔗 Related PRs
N/A

## 📝 Notes
N/A

Stefan-Ethernal committed:
fix: build and push both prod and dev images (#1157)

## 🔄 Changes Summary
During build and push Docker image, we haven't been handing the
development image (that includes shell), so this PR fixes it.

## ⚠️ Breaking Changes
N/A

## 📋 Config Updates
N/A

## ✅ Testing
- 🤖 **Automatic**: N/A
- 🖱️ **Manual**: Create new release and make sure that both prod Docker
image (without suffix) and the `development`
 (with dev suffix) are built.

## 🐞 Issues
- Closes #[issue-number]

## 🔗 Related PRs
- [Optional: Enumerate related pull requests]

## 📝 Notes
- [Optional: design decisions, tradeoffs, or TODOs]

adria0pol committed:
Merge branch 'develop' into fix/validate-external-signatures
